Libraries/LibJS/Runtime/ReflectObject.cpp

```diff
@@ -0,0 +1,247 @@
+}
+
+Value ReflectObject::define_property(Interpreter& interpreter)
+{
+    auto* target = get_target_object_from(interpreter, "defineProperty");
+    if (!target)
+        return {};
+    if (!interpreter.argument(2).is_object())
+        return interpreter.throw_exception<TypeError>("Descriptor argument is not an object");
+    auto property_key = interpreter.argument(1).to_string();
+    auto& descriptor = interpreter.argument(2).as_object();
+    auto success = target->define_property(property_key, descriptor, false);
+    return Value(success);
+}
+
+Value ReflectObject::delete_property(Interpreter& interpreter)
+{
+    auto* target = get_target_object_from(interpreter, "deleteProperty");
+    if (!target)
+        return {};
+
+    auto property_key = interpreter.argument(1);
+    if (property_key.to_number().is_finite_number() && (i32)property_key.to_double() == property_key.to_i32() && property_key.to_i32() >= 0)
+        return target->delete_property(PropertyName(property_key.to_i32()));
+    return target->delete_property(PropertyName(property_key.to_string()));
+}
+
+Value ReflectObject::get(Interpreter& interpreter)
+{
+    // FIXME: There's a third argument, receiver, for getters - use it once we have those.
+    auto* target = get_target_object_from(interpreter, "get");
+    if (!target)
+        return {};
+    auto property_key = interpreter.argument(1).to_string();
+    return target->get(property_key).value_or(js_undefined());
+}
+
+Value ReflectObject::get_own_property_descriptor(Interpreter& interpreter)
+{
+    auto* target = get_target_object_from(interpreter, "getOwnPropertyDescriptor");
+    if (!target)
+        return {};
+    auto property_key = interpreter.argument(1).to_string();
+    return target->get_own_property_descriptor(property_key);
+}
+
+Value ReflectObject::get_prototype_of(Interpreter& interpreter)
+{
+    auto* target = get_target_object_from(interpreter, "getPrototypeOf");
+    if (!target)
+        return {};
+    return target->prototype();
+}
+
+Value ReflectObject::has(Interpreter& interpreter)
+{
+    auto* target = get_target_object_from(interpreter, "has");
+    if (!target)
+        return {};
+    auto property_key = interpreter.argument(1).to_string();
+    return Value(target->has_property(property_key));
+}
+
+Value ReflectObject::is_extensible(Interpreter&)
+{
+    // FIXME: For this to be useful we need one of these:
+    // Object.seal(), Object.freeze(), Reflect.preventExtensions()
+    // For now we just return true, as that's always the case.
+    return Value(true);
+}
+
+Value ReflectObject::own_keys(Interpreter& interpreter)
+{
+    auto* target = get_target_object_from(interpreter, "ownKeys");
+    if (!target)
+        return {};
+    return target->get_own_properties(*target, GetOwnPropertyMode::Key);
+}
+
+Value ReflectObject::prevent_extensions(Interpreter&)
+{
+    // FIXME: Implement me :^)
+    ASSERT_NOT_REACHED();
+}
```

Comment:
This is probably as simple as adding a flag to the `JS::Object` class, but I'll leave that for another day.

Reply:
I think we can transition to a different `Shapé` and share the shape between all otherwise similar objects that have done a `preventExtensions()` :)